src/Evolution/DiscontinuousGalerkin/SlopeLimiters/Hierarchical.hpp

```diff
@@ -0,0 +1,982 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <boost/functional/hash.hpp>  // IWYU pragma: keep
+#include <cmath>
+#include <cstddef>
+#include <limits>
+#include <ostream>
+#include <pup.h>  // IWYU pragma: keep
+#include <string>
+#include <type_traits>
+#include <unordered_map>
+#include <utility>  // for pair
+#include <vector>
+
+#include "DataStructures/DataBox/DataBoxTag.hpp"
+#include "DataStructures/DataVector.hpp"  // IWYU pragma: keep
+#include "DataStructures/Index.hpp"
+#include "DataStructures/ModalVector.hpp"           // IWYU pragma: keep
+#include "DataStructures/Tensor/Metafunctions.hpp"  // IWYU pragma: keep
+#include "DataStructures/Tensor/Tensor.hpp"         // IWYU pragma: keep
+#include "DataStructures/Variables.hpp"             // IWYU pragma: keep
+#include "DataStructures/VariablesHelpers.hpp"
+#include "Domain/Direction.hpp"
+#include "Domain/Element.hpp"    // IWYU pragma: keep
+#include "Domain/ElementId.hpp"  // IWYU pragma: keep
+#include "Domain/Mesh.hpp"
+#include "Domain/OrientationMap.hpp"  // IWYU pragma: keep
+#include "Domain/Tags.hpp"            // IWYU pragma: keep
+#include "ErrorHandling/Assert.hpp"
+#include "ErrorHandling/Error.hpp"
+#include "NumericalAlgorithms/LinearOperators/CoefficientTransforms.hpp"
+#include "NumericalAlgorithms/Spectral/Spectral.hpp"
+#include "Options/Options.hpp"
+#include "Utilities/Algorithm.hpp"
+#include "Utilities/ForceInline.hpp"
+#include "Utilities/Gsl.hpp"
+#include "Utilities/Math.hpp"
+#include "Utilities/TMPL.hpp"
+
+// IWYU pragma: no_include <algorithm>
+// IWYU pragma: no_forward_declare Variables
+
+/// \cond
+namespace SlopeLimiters {
+template <size_t VolumeDim, typename TagsToLimit>
+class Hierarchical;
+}  // namespace SlopeLimiters
+/// \endcond
+
+namespace SlopeLimiters {
+enum class HierarchicalType { Krivodonova };
+
+std::ostream& operator<<(std::ostream& os, HierarchicalType t) noexcept;
+
+/// Given the `NodalTag` swap the `DataVector` with a `ModalVector`.
+template <typename NodalTag>
+struct ModalTag : db::SimpleTag {
+  using type =
+      TensorMetafunctions::swap_type<ModalVector, typename NodalTag::type>;
+  static std::string name() noexcept {
+    return "coefficients(" + NodalTag::name() + ")";
+  }
+};
+
+/*!
+ * \ingroup SlopeLimitersGroup
+ * \brief A general hierarchical slope limiter implementation.
+ *
+ * Implements hierarchical slope limiters that limit the logical derivatives of
+ * the `Tags` starting at the highest derivatives. The simplest of these
+ * limiters is described in Krivodonova 2007 \cite Krivodonova2007. The
```

Comment:
I think better remove author/year, it is available through the `\cite`

Reply:
No, she has multiple papers on this subject.

Reply:
In the documentation the bibliographical information is available on-hover, so I'd prefer to reduce redundancy, e.g. `is described in \cite Krivodonova2007`. That's the same information in the source code, and it's nicely rendered in the HTML.

Reply:
I still don't agree with you

Reply:
okay, I've done this, it's in a new fixup